### PR TITLE
Timeline refactoring

### PR DESCRIPTION
**Pull request by ganfra**
Some refactoring to the timeline, in the the optic of exposing interface only, and introduce tests.

### CODE REVIEW

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/StateEventRedactionChecker.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+import org.matrix.androidsdk.rest.model.MatrixError;
+import org.matrix.androidsdk.rest.model.RoomMember;
+import org.matrix.androidsdk.util.Log;
+
+import java.util.List;
+
+import javax.annotation.Nonnull;
+
+
+class StateEventRedactionChecker {
```
Comment (bmarty) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/StateEventRedactionChecker.java]: Javadoc

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelineEventListeners.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    /**
+     * Add an events listener.
+     *
+     * @param listener the listener to add.
+     */
+    public void add(@Nullable final EventTimeline.Listener listener) {
+        if (listener != null) {
+            synchronized (this) {
+                if (mListeners.indexOf(listener) == -1) {
```
Comment (bmarty) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelineEventListeners.java]: Use `contains()`

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelineInvitedRoomSyncHandler.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                                   @Nullable final InvitedRoomSync invitedRoomSync) {
+        mRoom = room;
+        mLiveEventHandler = liveEventHandler;
+        mInvitedRoomSync = invitedRoomSync;
+    }
+
+    /**
+     * Handle the invitation room events
+     */
+    public void handle() {
```
Comment (bmarty) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelineInvitedRoomSyncHandler.java]: It's strange that `InvitedRoomSync` is not passed as a param here, but in the constructor.

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelineJoinRoomSyncHandler.java [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+                                final boolean isGlobalInitialSync) {
+        mEventTimeline = eventTimeline;
+        mRoomSync = roomSync;
+        mTimelineStateHolder = timelineStateHolder;
+        mTimelineLiveEventHandler = timelineLiveEventHandler;
+        mIsGlobalInitialSync = isGlobalInitialSync;
+    }
+
+
+    public void handle() {
```
Comment (bmarty) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelineJoinRoomSyncHandler.java]: Javadoc + same remark than for TimelineInvitedRoomSyncHandler

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelineLiveEventHandler.java [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+import org.matrix.androidsdk.util.EventDisplay;
+import org.matrix.androidsdk.util.JsonUtils;
+import org.matrix.androidsdk.util.Log;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.annotation.Nonnull;
+
+class TimelineLiveEventHandler {
```
Comment (bmarty) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelineLiveEventHandler.java]: Javadoc

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelineLiveEventHandler.java [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+     * Store a live room event.
+     *
+     * @param event                   The event to be stored.
+     * @param checkRedactedStateEvent true to check if this event redacts a state event
+     */
+    private void storeLiveRoomEvent(@NonNull final MXDataHandler dataHandler,
+                                    @NonNull final IMXStore store,
+                                    @NonNull Event event,
+                                    final boolean checkRedactedStateEvent) {
+
```
Comment (bmarty) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelineLiveEventHandler.java]: extra space

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelinePushWorker.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+import org.matrix.androidsdk.MXDataHandler;
+import org.matrix.androidsdk.call.MXCall;
+import org.matrix.androidsdk.data.RoomState;
+import org.matrix.androidsdk.rest.model.Event;
+import org.matrix.androidsdk.rest.model.bingrules.BingRule;
+import org.matrix.androidsdk.util.BingRulesManager;
+import org.matrix.androidsdk.util.Log;
+
+class TimelinePushWorker {
```
Comment (bmarty) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelinePushWorker.java]: Javadoc

File: matrix-sdk/src/main/java/org/matrix/androidsdk/fragments/MatrixMessageListFragment.java [hunk 8]
```diff
@@ -516,22 +517,17 @@ public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
             // display a room preview
             else if (PREVIEW_MODE_READ_ONLY.equals(previewMode)) {
                 mAdapter.setIsPreviewMode(true);
-                mEventTimeLine = new EventTimeline(mSession.getDataHandler(), mRoomId);
+                mEventTimeLine = EventTimelineFactory.pastTimeline(mSession.getDataHandler(), mRoomId);
```
Comment (bmarty) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/fragments/MatrixMessageListFragment.java]: It does not look like a past Timeline

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelineInvitedRoomSyncHandler.java (hunk 3 above)
Comment (ganfra) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelineInvitedRoomSyncHandler.java]: It's a common pattern for breaking dependencies. So, it's just the first step for introducing tests and then refactor !

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelineEventListeners.java (hunk 2 above)
Comment (ganfra) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelineEventListeners.java]: Yeah, probably need some rework, just copy/paste at the moment

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java [hunk 9]
```diff
@@ -426,7 +410,7 @@ public void cancelRemoteHistoryRequest() {
     //================================================================================
 
     public String getRoomId() {
-        return mLiveTimeline.getState().roomId;
+        return mTimeline.getState().roomId;
```
[the remark this answers is not in the document]
Comment (ganfra) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java]: It's the same, but yeah, why not

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/EventTimeline.java [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+     * @return true if this timeline is the live one
+     */
+    boolean isLiveTimeline();
+
+    /**
+     * Get whether we are at the end of the message stream
+     *
+     * @return true if end has been reached
+     */
+    boolean hasReachedHomeServerForwardsPaginationEnd();
```
[the remark this answers is not in the document]
Comment (ganfra) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/EventTimeline.java]: Not sure, just copied/pasted

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java [hunk 11]
```diff
@@ -236,7 +220,7 @@ public boolean isOngoingConferenceCall() {
      */
     public void setIsLeft(boolean isLeft) {
         mIsLeft = isLeft;
```
[the remark this answers is not in the document]
Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java]: I think no...

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/MXEventTimeline.java [hunk 12]
```diff
@@ -0,0 +1,990 @@
...
+
+    /**
+     * the server provides a token even for the first room message (which should never change it is the creator message).
+     * so requestHistory always triggers a remote request which returns an empty json.
+     * try to avoid such behaviour
+     */
+    private String mBackwardTopToken = "not yet found";
+
+    // true when the current timeline is an historical one
+    private boolean mIsHistorical;
```
[the remark this answers is not in the document]
Comment (ganfra) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/MXEventTimeline.java]: Not sure, I think it's used when a room is left

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/EventTimeline.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/EventTimeline.java]: ok, no, looks ok

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/EventTimeline.java [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+     *
+     * @param eventCount    number of events we want to retrieve
+     * @param useCachedOnly to use the cached events list only (i.e no request will be triggered)
+     * @param callback      callback to implement to be informed that the pagination request has been completed. Can be null.
+     * @return true if request starts
+     */
+    boolean backPaginate(int eventCount, boolean useCachedOnly, ApiCallback<Integer> callback);
+
+    /**
+     * Request older messages.
```
[the remark this answers is not in the document]
Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/EventTimeline.java]: done

File: matrix-sdk/src/main/java/org/matrix/androidsdk/fragments/MatrixMessageListFragment.java (hunk 8 above)
Comment (ganfra) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/fragments/MatrixMessageListFragment.java]: Yes but it does have the features of a past timeline = InMemory Store and fake room...
Probably need some rework then.

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/EventTimeline.java [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+     * @param direction the direction.
+     * @param callback  the callback.
+     * @return true if the operation succeeds
+     */
+    boolean paginate(Direction direction, ApiCallback<Integer> callback);
+
+    /**
+     * Cancel any pending pagination requests
+     */
+    void cancelPaginationRequest();
```
[the remark this answers is not in the document]
Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/EventTimeline.java]: done

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/EventTimelineFactory.java [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+package org.matrix.androidsdk.data.timeline;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import org.matrix.androidsdk.MXDataHandler;
+import org.matrix.androidsdk.data.Room;
+import org.matrix.androidsdk.data.store.MXMemoryStore;
+
+public class EventTimelineFactory {
```
[the remark this answers is not in the document]
Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/EventTimelineFactory.java]: done

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/MXEventTimeline.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/MXEventTimeline.java]: ok, no, it's rather referring to a room user has left...

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/StateEventRedactionChecker.java (hunk 1 above)
Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/StateEventRedactionChecker.java]: done

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelineEventListeners.java (hunk 2 above)
Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelineEventListeners.java]: done

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelineInvitedRoomSyncHandler.java (hunk 3 above)
Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelineInvitedRoomSyncHandler.java]: ok

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelinePushWorker.java (hunk 7 above)
Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/TimelinePushWorker.java]: done

File: matrix-sdk/src/main/java/org/matrix/androidsdk/fragments/MatrixMessageListFragment.java (hunk 8 above)
Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/fragments/MatrixMessageListFragment.java]: ok, I've renamed the method

**Pull request merged**